compiler/src/iree/compiler/ConstEval/Runtime.cpp

```diff
@@ -427,37 +446,58 @@ TypedAttr CompiledBinary::convertVariantToAttribute(Location loc,
   return {};
 }
 
-void CompiledBinary::initialize(void *data, size_t length) {
+LogicalResult CompiledBinary::initialize(Location loc, void *data,
+                                         size_t length) {
   Runtime &runtime = Runtime::getInstance();
+  // Keep the sticky initStatus alive then free in |runtime|'s destructor.
+  if (failed(
+          handleRuntimeError(loc, runtime.initStatus, /*freeStatus=*/false))) {
+    return failure();
+  }
+
+  iree_status_t status = iree_ok_status();
 
   // Create driver and device.
   iree_hal_driver_t *driver = nullptr;
-  IREE_CHECK_OK(iree_hal_driver_registry_try_create(
-      runtime.registry, iree_make_cstring_view("local-task"),
-      iree_allocator_system(), &driver));
-  IREE_CHECK_OK(iree_hal_driver_create_default_device(
-      driver, iree_allocator_system(), &device));
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_driver_registry_try_create(
+        runtime.registry, iree_make_cstring_view("local-task"),
+        iree_allocator_system(), &driver);
+  }
+
+  if (iree_status_is_ok(status)) {
+    iree_hal_driver_create_default_device(driver, iree_allocator_system(),
+                                          &device);
+  }
   iree_hal_driver_release(driver);
 
   // Create hal module.
-  iree_hal_device_t *device_ptr = device.get();
-  IREE_CHECK_OK(iree_hal_module_create(
-      runtime.instance.get(), /*device_count=*/1, &device_ptr,
-      IREE_HAL_MODULE_FLAG_NONE, iree_allocator_system(), &hal_module));
+  if (iree_status_is_ok(status)) {
+    iree_hal_device_t *device_ptr = device.get();
```

Comment:
would be clearer if you did what's done below for modules (std::array and then passing array data()/size())